File: php/rio.php
<?php
    class Coffee {
        private $_water = 0.0;
        private $_powder = 0.0;
        public function addWater($quantity) {
            $this->_water += $quantity;
        }
        public function addPowder($quantity) {
            $this->_powder += $quantity;
        }
        public function getTotal() {
            return $this->_water + $this->_powder;
        }
        public function taste($quantity) {
            $total = $this->getTotal();
            $this->_water -= $quantity * $this->_water / $total;
            $this->_powder -= $quantity * $this->_powder / $total;
        }
        public function getConsentration($asPercent) {
            return ($this->_powder / $this->getTotal()) * ($asPercent ? 100.0 : 1.0);
        }
    }

    $coffee = new Coffee();
    $acts = (int)trim(fgets(STDIN));

    for ($count = 0; $count < $acts; $count++) {
        $arr = split(" ", trim(fgets(STDIN)));
        $actType = (int)$arr[0];
        $quantity = (int)$arr[1];
        switch ($actType) {
            case 1:
                $coffee->addWater($quantity);
                break;
            case 2:
                $coffee->addPowder($quantity);
                break;
            case 3:
                $coffee->taste($quantity);
                break;
        }
    }
    print((int)$coffee->getConsentration(true));
?>

File: perl/rio.pl
use strict;
use warnings;

package Coffee; {
    my $powder;
    my $water;

    sub new() {
        my $class = shift;
        my $self = {
            water => 0.0,
            powder => 0.0,
        };
        bless $self, $class;
        return $self;
    }

    sub addWater {
        my $self = shift;
        my $quantity = $_[0];
        $self->{water} += $quantity;
    }

    sub addPowder {
        my $self = shift;
        my $quantity = $_[0];
        $self->{powder} += $quantity;
    }

    sub getTotal {
        my $self = shift;
        return $self->{powder} + $self->{water};
    }

    sub taste {
        my $self = shift;
        my $quantity = $_[0];
        my $total = $self->getTotal;
        $self->{powder} -= $quantity * $self->{powder} / $total;
        $self->{water} -= $quantity * $self->{water} / $total;
    }

    sub getConsentration {
        my $self = shift;
        my $asPercent = $_[0];
        return ($self->{powder} / $self->getTotal) * ($asPercent ? 100.0 : 1.0);
    }
}

my $coffee = Coffee->new();
my $str = <STDIN>;
chomp($str);
my $acts = $str + 0;
for my $cnt (1...$acts) {
    my $actLine = <STDIN>;
    chomp($actLine);
    my @actArr = split(" ", $actLine);
    my $actType = $actArr[0];
    my $quantity = $actArr[1];
    if ($actType == 1) {
        $coffee->addWater($quantity);
    }elsif ($actType == 2) {
        $coffee->addPowder($quantity);
    }elsif ($actType == 3) {
        $coffee->taste($quantity);
    }
}

printf("%d\n", $coffee->getConsentration(1));
